### ISSUE TITLE
Process died unexpectedly (btrbk v0.30.0) - Use of uninitialized value in concatenation (.) or string

### ISSUE THREAD

**Issue by coolintentions** [digint/btrbk#344]
Hello, 
I get 'process died unexpectedly' Exception when running the script. Not shure if it's bad configuration or bug.

Versions:
btrbk v0.30.0
btrfs-progs version 5.4.1-2
Ubuntu 20.04

Configuration:
```
snapshot_preserve_min   2d
snapshot_preserve          14d
target_preserve_min        no
target_preserve               20d 10w *m
snapshot_dir                   @snapshoots
transaction_log                /var/log/btrbk.log
timestamp_format           long

volume /media/btrfs_root
  target /media/system_data_drive/btrfs_backup/snapshoots
  subvolume @
  subvolume @home
```

Debug: sudo btrbk -l debug dryrun
```
btrbk command line client, version 0.30.0  (Sat Oct 17 11:37:53 2020)
Using configuration: /etc/btrbk/btrbk.conf
Initializing volume section: /media/btrfs_root
Resolving btrfs mount point for: /media/btrfs_root
### readlink -v -e /media/btrfs_root
Command execution successful
Real path for "/media/btrfs_root" is: /media/btrfs_root
### cat /proc/self/mountinfo
Command execution successful
Btrfs mount point for "/media/btrfs_root": /media/btrfs_root (mount_source=/dev/sdc3, subvolid=5)
### btrfs subvolume list -a -c -u -q -R /media/btrfs_root
Command execution successful
Parsed 15 total subvolumes for filesystem at: /media/btrfs_root
### btrfs subvolume list -a -r /media/btrfs_root
Command execution successful
Parsed 0 readonly subvolumes for filesystem at: /media/btrfs_root
### btrfs subvolume show --rootid=5 /media/btrfs_root
Command execution successful
Parsed 11 subvolume detail items: /media/btrfs_root
Found btrfs root: /media/btrfs_root
Initializing subvolume section: /media/btrfs_root/@
Resolving btrfs mount point for: /media/btrfs_root/@
### readlink -v -e /media/btrfs_root/@
Command execution successful
Real path for "/media/btrfs_root/@" is: /media/btrfs_root/@
Btrfs mount point for "/media/btrfs_root/@": /media/btrfs_root (mount_source=/dev/sdc3, subvolid=5)
Found "/media/btrfs_root/@" (id=263) in btrfs tree of: /media/btrfs_root
Resolving btrfs mount point for: /media/btrfs_root/@snapshoots
### readlink -v -e /media/btrfs_root/@snapshoots
Command execution successful
Real path for "/media/btrfs_root/@snapshoots" is: /media/btrfs_root/@snapshoots
Btrfs mount point for "/media/btrfs_root/@snapshoots": /media/btrfs_root (mount_source=/dev/sdc3, subvolid=5)
Initializing subvolume section: /media/btrfs_root/@home
Resolving btrfs mount point for: /media/btrfs_root/@home
### readlink -v -e /media/btrfs_root/@home
Command execution successful
Real path for "/media/btrfs_root/@home" is: /media/btrfs_root/@home
Btrfs mount point for "/media/btrfs_root/@home": /media/btrfs_root (mount_source=/dev/sdc3, subvolid=5)
Found "/media/btrfs_root/@home" (id=264) in btrfs tree of: /media/btrfs_root
Resolving btrfs mount point for: /media/btrfs_root/@snapshoots
Btrfs mount point for "/media/btrfs_root/@snapshoots": /media/btrfs_root (mount_source=/dev/sdc3, subvolid=5)
Initializing target section: /media/system_data_drive/btrfs_backup/snapshoots
Resolving btrfs mount point for: /media/system_data_drive/btrfs_backup/snapshoots
### readlink -v -e /media/system_data_drive/btrfs_backup/snapshoots
Command execution successful
Real path for "/media/system_data_drive/btrfs_backup/snapshoots" is: /media/system_data_drive/btrfs_backup/snapshoots
Btrfs mount point for "/media/system_data_drive/btrfs_backup/snapshoots": /media/system_data_drive/btrfs_backup (mount_source=/dev/sdb, subvolid=258)
### btrfs subvolume list -a -c -u -q -R /media/system_data_drive/btrfs_backup
Command execution successful
Parsed 1 total subvolumes for filesystem at: /media/system_data_drive/btrfs_backup
### btrfs subvolume list -a -r /media/system_data_drive/btrfs_backup
Command execution successful
Parsed 0 readonly subvolumes for filesystem at: /media/system_data_drive/btrfs_backup
### btrfs subvolume show --rootid=5 /media/system_data_drive/btrfs_backup
Command execution successful
Parsed 11 subvolume detail items: /media/system_data_drive/btrfs_backup
Found btrfs root: /media/system_data_drive/btrfs_backup

ERROR: process died unexpectedly (btrbk v0.30.0)
Please contact the author: Axel Burri <axel@tty0.ch>

Stack Trace:
----------------------------------------
Use of uninitialized value in concatenation (.) or string at /usr/bin/btrbk line 2367.
 at /usr/bin/btrbk line 315.
        main::__ANON__("Use of uninitialized value in concatenation (.) or string at "...) called at /usr/bin/btrbk line 2367
        main::btr_tree(HASH(0x5607dc14dca8), 258, "/dev/sdb", ARRAY(0x5607dc14da38)) called at /usr/bin/btrbk line 2709
        main::vinfo_init_root(HASH(0x5607dc112918)) called at /usr/bin/btrbk line 6168
```

**Comment by digint** [comment]
Thanks for reporting!

Yes, it's a bug: for some reason, btrbk can't assign mount points to btrfs subvolumes, and runs into a buggy line of code.

In order to dig deeper into this, could you please provide me with some info:

```
cat /proc/self/mountinfo
```
and
```
btrfs subvolume list -a -c -u -q -R /media/system_data_drive/btrfs_backup
```

while at it, you can try the fix from the fix_344 branch: (the fix in 81c92940c1645be9f9d87958e7f02395f0a15b39 should now give you some warning instead of dying)
```
cd /tmp/
wget https://raw.githubusercontent.com/digint/btrbk/fix_344/btrbk
chmod +x btrbk
sudo /tmp/btrbk dryrun
```

**Comment by coolintentions** [comment]
Hi, thanks for quick reply. It seems that fix works. :)

cat /proc/self/mountinfo
```
24 32 0:22 / /sys rw,nosuid,nodev,noexec,relatime shared:7 - sysfs sysfs rw
25 32 0:5 / /proc rw,nosuid,nodev,noexec,relatime shared:15 - proc proc rw
26 32 0:6 / /dev rw,nosuid,noexec,relatime shared:2 - devtmpfs udev rw,size=8051440k,nr_inodes=2012860,mode=755
27 26 0:23 / /dev/pts rw,nosuid,noexec,relatime shared:3 - devpts devpts rw,gid=5,mode=620,ptmxmode=000
28 32 0:24 / /run rw,nosuid,nodev,noexec,relatime shared:5 - tmpfs tmpfs rw,size=1619012k,mode=755
32 1 0:27 /@ / rw,noatime,nodiratime shared:1 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=263,subvol=/@
31 24 0:7 / /sys/kernel/security rw,nosuid,nodev,noexec,relatime shared:8 - securityfs securityfs rw
33 26 0:30 / /dev/shm rw,nosuid,nodev shared:4 - tmpfs tmpfs rw
34 28 0:31 / /run/lock rw,nosuid,nodev,noexec,relatime shared:6 - tmpfs tmpfs rw,size=5120k
35 24 0:32 / /sys/fs/cgroup ro,nosuid,nodev,noexec shared:9 - tmpfs tmpfs ro,mode=755
36 35 0:33 / /sys/fs/cgroup/unified rw,nosuid,nodev,noexec,relatime shared:10 - cgroup2 cgroup2 rw,nsdelegate
37 35 0:34 / /sys/fs/cgroup/systemd rw,nosuid,nodev,noexec,relatime shared:11 - cgroup cgroup rw,xattr,name=systemd
38 24 0:35 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime shared:12 - pstore pstore rw
39 24 0:36 / /sys/firmware/efi/efivars rw,nosuid,nodev,noexec,relatime shared:13 - efivarfs efivarfs rw
40 24 0:37 / /sys/fs/bpf rw,nosuid,nodev,noexec,relatime shared:14 - bpf none rw,mode=700
41 35 0:38 / /sys/fs/cgroup/cpu,cpuacct rw,nosuid,nodev,noexec,relatime shared:16 - cgroup cgroup rw,cpu,cpuacct
42 35 0:39 / /sys/fs/cgroup/rdma rw,nosuid,nodev,noexec,relatime shared:17 - cgroup cgroup rw,rdma
43 35 0:40 / /sys/fs/cgroup/memory rw,nosuid,nodev,noexec,relatime shared:18 - cgroup cgroup rw,memory
44 35 0:41 / /sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime shared:19 - cgroup cgroup rw,net_cls,net_prio
45 35 0:42 / /sys/fs/cgroup/freezer rw,nosuid,nodev,noexec,relatime shared:20 - cgroup cgroup rw,freezer
46 35 0:43 / /sys/fs/cgroup/perf_event rw,nosuid,nodev,noexec,relatime shared:21 - cgroup cgroup rw,perf_event
47 35 0:44 / /sys/fs/cgroup/cpuset rw,nosuid,nodev,noexec,relatime shared:22 - cgroup cgroup rw,cpuset
48 35 0:45 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime shared:23 - cgroup cgroup rw,hugetlb
49 35 0:46 / /sys/fs/cgroup/blkio rw,nosuid,nodev,noexec,relatime shared:24 - cgroup cgroup rw,blkio
50 35 0:47 / /sys/fs/cgroup/pids rw,nosuid,nodev,noexec,relatime shared:25 - cgroup cgroup rw,pids
51 35 0:48 / /sys/fs/cgroup/devices rw,nosuid,nodev,noexec,relatime shared:26 - cgroup cgroup rw,devices
52 25 0:49 / /proc/sys/fs/binfmt_misc rw,relatime shared:27 - autofs systemd-1 rw,fd=28,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=20035
53 26 0:20 / /dev/mqueue rw,nosuid,nodev,noexec,relatime shared:28 - mqueue mqueue rw
54 26 0:50 / /dev/hugepages rw,relatime shared:29 - hugetlbfs hugetlbfs rw,pagesize=2M
55 24 0:8 / /sys/kernel/debug rw,nosuid,nodev,noexec,relatime shared:30 - debugfs debugfs rw
56 24 0:12 / /sys/kernel/tracing rw,nosuid,nodev,noexec,relatime shared:31 - tracefs tracefs rw
57 24 0:51 / /sys/fs/fuse/connections rw,nosuid,nodev,noexec,relatime shared:32 - fusectl fusectl rw
58 24 0:21 / /sys/kernel/config rw,nosuid,nodev,noexec,relatime shared:33 - configfs configfs rw
129 32 0:27 /@home /home rw,noatime,nodiratime shared:69 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=264,subvol=/@home
132 32 0:27 / /media/btrfs_root rw,noatime,nodiratime shared:71 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=5,subvol=/
128 28 0:54 / /run/shm rw,nosuid,noexec,relatime shared:73 - tmpfs tmpfs rw
137 32 0:27 /@/media/raid5_data_drive/arsenic /smb_share/arsenic rw,noatime,nodiratime shared:1 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=263,subvol=/@
140 32 8:33 / /boot/efi rw,relatime shared:76 - vfat /dev/sdc1 rw,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro
143 32 0:27 /@/media/raid5_data_drive/medic /smb_share/medic rw,noatime,nodiratime shared:1 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=263,subvol=/@
146 32 0:27 /@/media/raid5_data_drive/susa /smb_share/susa rw,noatime,nodiratime shared:1 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=263,subvol=/@
149 32 0:27 /@/media/raid5_data_drive/bjelac /smb_share/bjelac rw,noatime,nodiratime shared:1 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=263,subvol=/@
153 32 0:27 /@snapshoots /snapshoots rw,noatime,nodiratime shared:81 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=266,subvol=/@snapshoots
152 32 7:0 / /snap/core18/1885 ro,nodev,relatime shared:83 - squashfs /dev/loop0 ro
158 32 7:1 / /snap/lxd/17629 ro,nodev,relatime shared:85 - squashfs /dev/loop1 ro
163 32 0:53 / /media/system_data_drive/btrfs_root rw,noatime,nodiratime shared:87 - btrfs /dev/sdb rw,compress=lzo,discard,space_cache,subvolid=5,subvol=/
162 32 0:53 /@btrfs_backup /media/system_data_drive/btrfs_backup rw,noatime,nodiratime shared:89 - btrfs /dev/sdb rw,compress=lzo,discard,space_cache,subvolid=258,subvol=/@btrfs_backup
161 32 7:2 / /snap/snapd/9279 ro,nodev,relatime shared:91 - squashfs /dev/loop2 ro
170 32 7:3 / /snap/snapd/9607 ro,nodev,relatime shared:93 - squashfs /dev/loop3 ro
173 32 7:4 / /snap/core18/1880 ro,nodev,relatime shared:95 - squashfs /dev/loop4 ro
176 32 7:5 / /snap/lxd/17738 ro,nodev,relatime shared:97 - squashfs /dev/loop5 ro
1189 28 0:61 / /run/user/997 rw,nosuid,nodev,relatime shared:650 - tmpfs tmpfs rw,size=1619008k,mode=700,uid=997,gid=997
1211 32 0:27 /@/var/lib/docker/btrfs /var/lib/docker/btrfs rw,noatime,nodiratime shared:1 - btrfs /dev/sdc3 rw,compress=lzo,ssd,discard,space_cache,subvolid=263,subvol=/@
1232 28 0:24 /snapd/ns /run/snapd/ns rw,nosuid,nodev,noexec,relatime - tmpfs tmpfs rw,size=1619012k,mode=755
1258 1232 0:4 mnt:[4026532502] /run/snapd/ns/lxd.mnt rw - nsfs nsfs rw
1259 28 0:63 / /run/user/116 rw,nosuid,nodev,relatime shared:682 - tmpfs tmpfs rw,size=1619008k,mode=700,uid=116,gid=126
1288 129 0:64 / /home/dusko/.cache/gvfs rw,nosuid,nodev,relatime shared:694 - fuse.gvfsd-fuse gvfsd-fuse rw,user_id=1000,group_id=1000
1321 1259 0:65 / /run/user/116/gvfs rw,nosuid,nodev,relatime shared:698 - fuse.gvfsd-fuse gvfsd-fuse rw,user_id=116,group_id=126
987 32 8:1 / /media/raid5_data_drive rw,relatime shared:537 - ext4 /dev/sda1 rw
961 28 0:60 / /run/user/1000 rw,nosuid,nodev,relatime shared:524 - tmpfs tmpfs rw,size=1619008k,mode=700,uid=1000,gid=1000
```

btrfs subvolume list -a -c -u -q -R /media/system_data_drive/btrfs_backup
`ID 258 gen 15 cgen 14 top level 5 parent_uuid -    received_uuid -     uuid 6c1203b1-732d-5142-8209-df6c34b70531 path <FS_TREE>/@btrfs_backup`


fix_344
sudo /tmp/btrbk dryrun
```
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.30.1-dev)

    Date:   Mon Oct 19 17:57:34 2020
    Config: /etc/btrbk/btrbk.conf
    Dryrun: YES

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/media/btrfs_root/@
+++ /media/btrfs_root/@snapshoots/@.20201019T1757
*** /media/system_data_drive/btrfs_backup/snapshoots/@.20201019T1757

/media/btrfs_root/@home
+++ /media/btrfs_root/@snapshoots/@home.20201019T1757
*** /media/system_data_drive/btrfs_backup/snapshoots/@home.20201019T1757

NOTE: Dryrun was active, none of the operations above were actually executed!
```

**Comment by digint** [comment]
Strange, I was expecting at least a warning when running the fixed btrbk, something like:
```
WARN "Unknown subvolid=$node_id (in btrfs tree of $host_mount_source) for mountpoint: $vol->{URL_PREFIX}$_->{mount_point}";
```

Probably something changed in your disk setup in the meantime. Let me know if you encounter this message again, I will leave this issue open anyways until v0.30.1 gets released.

**Comment by coolintentions** [comment]
I haven't changed anything since I found that bug. Didn't even rebooted my server. My drives are mounted over fstab.
It just works with the fix.

**Comment by digint** [comment]
Great, thanks.

Merged fix to master: 81c92940c1645be9f9d87958e7f02395f0a15b39, will close bug on next release

**Comment by digint** [comment]
Included in [v0.31.0](/digint/btrbk/releases/tag/v0.31.0)
